// angular-plotly.js/esm5/src/app/shared/plotly.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export var Plotly;
(function (Plotly) {
    /**
     * @record
     */
    function Figure() { }
    Plotly.Figure = Figure;
    if (false) {
        /** @type {?} */
        Figure.prototype.data;
        /** @type {?} */
        Figure.prototype.layout;
        /** @type {?} */
        Figure.prototype.frames;
    }
    /**
     * @record
     */
    function PlotlyHTMLElement() { }
    Plotly.PlotlyHTMLElement = PlotlyHTMLElement;
    if (false) {
        /**
         * @param {?} event
         * @param {?} callback
         * @return {?}
         */
        PlotlyHTMLElement.prototype.on = function (event, callback) { };
        /**
         * @param {?} event
         * @param {?} callback
         * @return {?}
         */
        PlotlyHTMLElement.prototype.removeListener = function (event, callback) { };
    }
    /**
     * @record
     */
    function PlotlyInstance() { }
    Plotly.PlotlyInstance = PlotlyInstance;
    if (false) {
        /** @type {?} */
        PlotlyInstance.prototype.Plots;
        /**
         * @param {?} div
         * @param {?} data
         * @param {?=} layout
         * @param {?=} config
         * @return {?}
         */
        PlotlyInstance.prototype.newPlot = function (div, data, layout, config) { };
        /**
         * @param {?} div
         * @param {?} data
         * @param {?=} layout
         * @param {?=} config
         * @return {?}
         */
        PlotlyInstance.prototype.plot = function (div, data, layout, config) { };
        /**
         * @param {?} div
         * @param {?} data
         * @param {?=} layout
         * @param {?=} config
         * @return {?}
         */
        PlotlyInstance.prototype.react = function (div, data, layout, config) { };
    }
})(Plotly || (Plotly = {}));
